helpers/embeds.js

```diff
@@ -0,0 +1,78 @@
+import { findComposedAncestor } from '../helpers/dom.js';
+import { html } from 'lit';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		for await (const childNode of this.#childNodes) {
+			await childNode.renderNode(options);
+		}
+
+		if (this.#node.id === '3') {
+			console.log('wee');
+		}
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) this.#node.replaceWith(embed);
+	}
+
+}
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	/* NOT IMPLEMENTED */
+}
+
+export async function renderEmbeds(node, options) {
+	const elems = [...node.querySelectorAll('div[data-d2l-embed-type], span[data-d2l-embed-type]')];
+	if (elems.length === 0) return;
+
+	const renderNodes = new Map();
+	const rootNodes = [];
+	const constructRenderNode = elem => {
+		if (renderNodes.has(elem)) return;
+
+		const renderNode = new RenderNode(elem);
+		renderNodes.set(elem, renderNode);
+
+		const nearestEmbedAncestor = findComposedAncestor(elem, ancestor => ancestor !== elem && elems.includes(ancestor));
```

Comment:
Possibly overkill since we won't actually be dealing with web components up the chain, but it was a quick and dirty way to find ancestors.

We're basically making use of `querySelectorAll` returning in DOM order here. If we don't find an ancestor that's in our list of embeds, then we're dealing with a root node. If we do, then we know it must have been processed already (and thus must be in the list of `renderNodes`) because it must have come earlier in the node list.

Reply:
Could this alternatively be:
```javascript
findComposedAncestor(elem, ancestor => ancestor !== elem && elem.hasAttribute('data-d2l-embed-type'));
```
The benefit being avoiding an array search for every element in the composed path.